### ISSUE TITLE
dynamices crm annotations

### ISSUE THREAD

**Issue by vladimirsakic** [object/Simple.OData.Client#411]
dynamics crm odata service returns annotations like this one:
{
 "@odata.count":1,
 "value":[
    {
      "@odata.etag":"W/\"1608905\"",
      "_dot_propertyid_value@OData.Community.Display.V1.FormattedValue":"Address Boulevard",
      "_dot_propertyid_value":"9a11328b-5c15-e711-80e5-1458d043b558",
    }
  ]
}

I have field _dot_propertyid_value, but how can I map to _dot_propertyid_value@OData.Community.Display.V1.FormattedValue

Thanks

**Comment by object** [comment]
Hi,
Currently the library doesn't support handling of custom annotations (it only supports standard annotations defined by OData protocol: id, count, deltaLink, nextPageLink etc.) So I am afraid it's not possible to achieve what you want without intercepting the response using custom rewriter.

**Comment by vladimirsakic** [comment]
Maybe it's good idea to write that rewriter for dynamics crm, because as I know, most developers use this library for communicating with web api endpoint

**Comment by object** [comment]
That can make sense. Unfortunately I have never used Dynamics CRM, so I have no knowledge of what kind of annotations such rewriter should be aware of.

**Comment by vladimirsakic** [comment]
I can send you examples if you are interested. Basicly it's just to receve string name of related entity (besides primary key) so you dont have to use expand if thats only field you need from related entity. And it's usefull because you can expand only one level so it saves you from looping and quering if you need two level expand for related entity name


On Jul 10, 2017, at 17:09, Vagif Abilov <notifications@github.com<mailto:notifications@github.com>> wrote:


That can make sense. Unfortunately I have never used Dynamics CRM, so I have no knowledge of what kind of annotations such rewriter should be aware of.

**Comment by object** [comment]
This sounds like a candidate for the next major version upgrade. And it has to be opted in in a smart way, i.e. developers not using Dynamic CRM shouldn't be aware of that. But if you have a reference to documentated annotations as well as some examples, it's worth taking into account.

**Comment by vladimirsakic** [comment]
As far as I know @OData.Community.Display.V1.FormattedValue is only field level annotation, as I sad, purpose id to get some string value along with guid or int id. Header to turn on this annotations is "Prefer: odata.include-annotations=OData.Community.Display.V1.FormattedValue"

https://msdn.microsoft.com/en-us/library/gg334767.aspx#bkmk_includeFormattedValues

**Comment by vladimirsakic** [comment]
Sorry, there's two more: @Microsoft.Dynamics.CRM.associatednavigationproperty and @Microsoft.Dynamics.CRM.lookuplogicalname

**Comment by vladimirsakic** [comment]
and @odata.nextLink

**Comment by object** [comment]
But this one is a part of OData spec.

**Comment by object** [comment]
So far no progress, but I see this request is coming from several people, so I plan to start working on it.

**Comment by bdebaere** [comment]
@object Isn't this already implemented seeing as https://github.com/OData/vocabularies/blob/master/OData.Community.Display.V1.md speaks of using it for property values? I can't find an example online on how to code it though.

**Comment by toryb** [comment]
@object I need to get field-level annotations for Dynamics CRM. 

Could you provide a little more information about your suggestion earlier in this thread that someone could "intercept the response using custom rewriter"? 

I assume you mean to assign an action to ODataClientSettings.AfterResponse in order to intercept the response. If I do that, what I am going to do with the annotations to surface them in the value returned by FindEntriesAsync(annotations)? Any direction would be helpful.

Thanks.

**Comment by flieks** [comment]
@toryb I also can't find any examples on how to implement the odata.include-annotations defined in the [odata web api docs](http://odata.github.io/WebApi/#04-19-odataIncludeAnnotations)

@vladimirsakic did you find a solution?

@object ODataClientSettings.AfterResponse is for the client side right ? But the server has to include other values in the response based on Display attributes.

**Comment by toryb** [comment]
@flieks You are correct, the server does need to include annotations. This is done based on the request to include annotations in the [prefer header](http://docs.oasis-open.org/odata/odata/v4.0/os/part1-protocol/odata-v4.0-os-part1-protocol.html#_Toc372793628) of the request (i.e. Prefer: odata.include-annotations="*"). I have used the ODataClientSettings.BeforeResponse to add the request to the header so I get all annotations with each request.

@vladimirsakic The way I have implemented the handling of CRM annotations is in the ODataClientSettings.AfterResponse. The way I handle them is to parse the content into a JObject and find the properties that contain an "@" (but don't start with "@"). Here is that line of code:

`var annotationProperties = contentObject.Properties().Where(x => x.Name.Contains("@") && !x.Name.StartsWith("@"));`

Then, by splitting the property name by '@' I get the property it applies to and the annotation to be applied:

     foreach (var prop in annotationProperties)
     {
        // Get the field name
        var values = prop.Name.Split(new char[] { '@' }, StringSplitOptions.RemoveEmptyEntries);
        if (values.Length != 2) break;
        if (Entity.FieldDictionary.ContainsKey(values[0]))
        {
          var entityField = Entity[values[0]];
          string value = prop.Value.ToString();
          string annotationType = values[1];
          switch (annotationType)
          {
            case "OData.Community.Display.V1.FormattedValue":
              {
                PropertyInfo pi = entityField.FieldType.GetProperty("FormattedValue");
                pi.SetValue(Entity[values[0]].FieldImpl, value);
              }
              break;
            case "Microsoft.Dynamics.CRM.lookuplogicalname":
              {
                PropertyInfo pi = entityField.FieldType.GetProperty("LookupLogicalName");
                // System.Diagnostics.Debug.Assert(prop.Value is string);
                pi.SetValue(Entity[values[0]].FieldImpl, value);
              }
              break;
            case "Microsoft.Dynamics.CRM.associatednavigationproperty":
              {
                PropertyInfo pi = entityField.FieldType.GetProperty("AssociatedNavigationProperty");
                // System.Diagnostics.Debug.Assert(prop.Value is string);
                pi.SetValue(Entity[values[0]].FieldImpl, value);
              }
              break;
          }
        } // end switch annotation type
      } // end foreach

There is probably much in there that needs explanation, but the essence of it is to search the response for field-level annotations and apply them to your implementation of the object as appropriate.

I hope that helps.

**Comment by EduardoRomo** [comment]
Any updates on this feature?

**Comment by Soulfire86** [comment]
Needing this here in 2024. Any ideas?
